momaland/envs/multiwalker_stability/momultiwalker_stability_base.py

```diff
@@ -0,0 +1,228 @@
+"""Adapted from the Multiwalker problem.
+
+From Gupta, J. K., Egorov, M., and Kochenderfer, M. (2017). Cooperative multi-agent control using
+deep reinforcement learning. International Conference on Autonomous Agents and Multiagent Systems
+"""
+
+from typing_extensions import override
+
+import numpy as np
+from gymnasium import spaces
+from pettingzoo.sisl.multiwalker.multiwalker_base import (
+    FPS,
+    LEG_H,
+    SCALE,
+    TERRAIN_GRASS,
+    TERRAIN_HEIGHT,
+    TERRAIN_LENGTH,
+    TERRAIN_STARTPAD,
+    TERRAIN_STEP,
+    VIEWPORT_W,
+    WALKER_SEPERATION,
+)
+from pettingzoo.sisl.multiwalker.multiwalker_base import (
+    BipedalWalker as pz_bipedalwalker,
+)
+from pettingzoo.sisl.multiwalker.multiwalker_base import (
+    MultiWalkerEnv as pz_multiwalker_base,
+)
+
+
+class MOBipedalWalkerStability(pz_bipedalwalker):
+    """Walker Object with the physics implemented."""
+
+    @override
+    def __init__(
+        self,
+        world,
+        forward_reward,
+        fall_reward,
+        terminate_reward,
+        stability_reward,
+        init_x=TERRAIN_STEP * TERRAIN_STARTPAD / 2,
+        init_y=TERRAIN_HEIGHT + 2 * LEG_H,
+        n_walkers=2,
+        seed=None,
+        terrain_length=TERRAIN_LENGTH,
+        terrain_step=TERRAIN_STEP,
+    ):
+        super().__init__(world, init_x, init_y, n_walkers, seed)
+        self.forward_reward = forward_reward
+        self.fall_reward = fall_reward
+        self.terminate_reward = terminate_reward
+        self.stability_reward = stability_reward
+        self.terrain_length = terrain_length
+        self.terrain_step = terrain_step
+
+    @property
+    def reward_space(self):
+        """Reward space shape = 2 element 1D array, each element representing 1 objective.
+
+        1. scalarized reward of: package moving forward, no walkers falling, and package not falling
+        2. package not tipping
+        """
+        return spaces.Box(
+            low=np.array([-210, -1.75]),
+            high=np.array([0.46, 0]),
+            shape=(2,),
+            dtype=np.float32,
+        )
+
+
+class MOMultiWalkerStabilityEnv(pz_multiwalker_base):
+    """Multiwalker problem domain environment engine.
+
+    Deals with the simulation of the environment.
+    """
+
+    @override
+    def __init__(
+        self,
+        n_walkers=3,
+        position_noise=1e-3,
+        angle_noise=1e-3,
+        forward_reward=1.0,
+        terminate_reward=-100.0,
+        fall_reward=-10.0,
+        stability_reward=-100,
+        shared_reward=True,
+        terminate_on_fall=True,
+        remove_on_fall=True,
+        terrain_length=TERRAIN_LENGTH,
+        max_cycles=500,
+        render_mode=None,
+    ):
+        self.stability_reward = stability_reward
+        self.previous_pkg_angle = 0
```

Comment:
Why not `self.package.angle` after init?

Reply:
`self.package` is created in `reset()`. I can move it to the end of reset.